// color-thief-react/lib/utils/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadImage = exports.getPredominantColorFromImgURL = exports.rgbStringfy = exports.getColorsPaletteFromImgUrl = exports.formatRGB = exports.reducer = exports.initialReducerState = void 0;
var initialReducerState_1 = require("./initialReducerState");
Object.defineProperty(exports, "initialReducerState", { enumerable: true, get: function () { return __importDefault(initialReducerState_1).default; } });
var reducer_1 = require("./reducer");
Object.defineProperty(exports, "reducer", { enumerable: true, get: function () { return __importDefault(reducer_1).default; } });
var formatRGB_1 = require("./formatRGB");
Object.defineProperty(exports, "formatRGB", { enumerable: true, get: function () { return __importDefault(formatRGB_1).default; } });
var getColorsPaletteFromImgUrl_1 = require("./getColorsPaletteFromImgUrl");
Object.defineProperty(exports, "getColorsPaletteFromImgUrl", { enumerable: true, get: function () { return __importDefault(getColorsPaletteFromImgUrl_1).default; } });
var rgbStringfy_1 = require("./rgbStringfy");
Object.defineProperty(exports, "rgbStringfy", { enumerable: true, get: function () { return __importDefault(rgbStringfy_1).default; } });
var getPredominantColorFromImgURL_1 = require("./getPredominantColorFromImgURL");
Object.defineProperty(exports, "getPredominantColorFromImgURL", { enumerable: true, get: function () { return __importDefault(getPredominantColorFromImgURL_1).default; } });
var loadImage_1 = require("./loadImage");
Object.defineProperty(exports, "loadImage", { enumerable: true, get: function () { return __importDefault(loadImage_1).default; } });
